02 Charging For Your App/app.rb

```diff
@@ -0,0 +1,207 @@
+require 'shopify_api'
+require 'sinatra'
+require 'httparty'
+require 'dotenv'
+
+
+class GiftBasket < Sinatra::Base
+  attr_reader :key,:secret,:tokens
+
+  def initialize
+    Dotenv.load
+    @key = ENV['API_KEY']
+    @secret = ENV['API_SECRET']
+    @app_url = "jamied.ngrok.io"
+    @tokens = {}
+    super
+  end
+
+  get '/giftbasket/install' do
+    shop = request.params['shop']
+    scopes = "read_orders,read_products,write_products"
+
+    # construct the installation URL and redirect the merchant
+    install_url = "http://#{shop}/admin/oauth/authorize?client_id=#{@key}"\
+                "&scope=#{scopes}&redirect_uri=https://#{@app_url}/giftbasket/auth"
+
+    # redirect to the install_url
+    redirect install_url
+  end
+
+  get '/giftbasket/auth' do
+    # extract shop data from request parameters
+    shop = request.params['shop']
+    code = request.params['code']
+    hmac = request.params['hmac']
+
+    # perform hmac validation to determine if the request is coming from Shopify
+    validate_hmac(hmac,request)
+
+    # if no access token for this particular shop exist,
+    # POST the OAuth request and receive the token in the response
+    get_shop_access_token(shop,key,secret,code)
+
+    # now that the session is activated, create a recurring application charge
+    create_recurring_application_charge
+
+    # redirect to the bulk edit URL if there is a token and an activated session and an activated RecurringApplicationCharge
+    redirect bulk_edit_url
+  end
+
+  get '/activatecharge' do
+    # store the charge_id from the request
+    charge_id  = request.params['charge_id']
+    recurring_application_charge = ShopifyAPI::RecurringApplicationCharge.find(charge_id)
+    recurring_application_charge.status == "accepted" ? recurring_application_charge.activate : redirect(@tokens[:confirmation_url])
+
+    # once the charge is activated, subscribe to the order/create webhook and redirect the user back to the bulk edit URL
+    create_order_webhook
+    redirect bulk_edit_url
+  end
+
+  post '/giftbasket/webhook/order_create' do
+    # inspect hmac value in header and verify webhook
+    hmac = request.env['HTTP_X_SHOPIFY_HMAC_SHA256']
+
+    request.body.rewind
+    data = request.body.read
+    webhook_ok = verify_webhook(hmac, data)
+
+    if webhook_ok
+      shop = request.env['HTTP_X_SHOPIFY_SHOP_DOMAIN']
+      token = @tokens[shop]
+
+      if not token.nil?
+        session = ShopifyAPI::Session.new(shop, token)
+        ShopifyAPI::Base.activate_session(session)
+      else
+        return [403, "You're not authorized to perform this action."]
+      end
+    else
+      return [403, "You're not authorized to perform this action."]
+    end
+
+    create_usage_charge
```

Comment:
this is entirely semantical, but it might make more sense to create the `usage_charge`  when the ingredients key matches. bonus points for checking the namespace as well.

this way the merchant only gets charged when the order contains a gift basket instead of being charged for every order they process 😄